Automate creation of remote repo

caldwellst opened the pull request:
Addresses #215. Tested with a fake organisation I made `nesta-cookiecutter-test`. I can create a public or private repo and it transfers. 

Worried from the documentation that it requires repo permissions anyway to transfer to a repository (e.g. you can't propose a transfer for approval), but would be good to test that. Need to fix one thing where the shell views a return JSON from one of the calls. Have ignored testing for now, need to check how to safely run `post_gen_project.sh` when testing (also maybe mocking of the `gh` API to confirm correct calls made?).

---

Checklist:

- [x] Updated documentation
- [x] CI passes
- [ ] Labelled PR major/minor/patch

caldwellst committed:
Fix gh CLI calls

caldwellst committed:
Use pwd instead of project slug

caldwellst committed:
Fix repo_name call in project cleanup

caldwellst committed:
Add initial update to post_gen_project.sh for testing

caldwellst committed:
Improve messaging

caldwellst committed:
Again fix the repo name

HarrisonWilde commented:
Interesting idea to solve it

caldwellst commented:
Hopefully redundant if we can just directly create repositories.

sqr00t commented:
There was a time when we created repos automatically, and it ended up creating too many so we stepped back from it. The tech team handles repo creation and other admin.

sqr00t commented on hooks/post_gen_project.sh:
```diff
@@ -17,7 +20,29 @@ git add .
+
+ISSUE_URL=$(gh issue create --repo nesta-cookiecutter-test/github_support --title "$ISSUE_TITLE" --body "$ISSUE_BODY")
+
+echo "Repository created and transferred: https://github.com/nesta-cookiecutter-test/$REPO_NAME"
```
Might want to check with the tech team on what the permissions would be to create repos in the Nesta org

caldwellst committed:
Simplify repo creation process

caldwellst committed:
Change messages

caldwellst committed:
Add tests

caldwellst committed:
Merge with master

caldwellst committed:
Get tests to pass

caldwellst committed:
Update version

caldwellst commented:
Updated based on #257. Main changes are that I removed the `autosetup` argument, and instead simplified it to simply `create_repo`. It will autosetup precommit, venv, etc. but requires the user to confirm they want to create a remote repository.

Updated docs to reflect this and added tests. Bumped version to 0.8.2.

HarrisonWilde commented on hooks/post_gen_project.sh:
```diff
@@ -181,29 +181,22 @@ git add .
+{% if cookiecutter.create_remote == 'yes' %}
+# Verify gh auth before touching remote
+if ! gh auth status >/dev/null 2>&1; then
+    echo "Error: gh not authenticated. Run 'gh auth login' then push manually." >&2
```
I wonder if for ease we should also just link to the docs and tell them to follow the manual step 3?

Reason being it isn't really just about then "pushing manually" right, they'd want to potentially also create the repo with the appropriate flags and so on.

The other option might be to check `gh auth status` right at the start (when the `create_repo` flag is true, and exit 0 if not encouraging the user to do this before anything actually happens in the cookie cutter?

HarrisonWilde commented on cookiecutter.json:
```diff
@@ -18,7 +18,6 @@
     "file_structure": "Select a file structure",
-    "use_r": "Use R?",
-    "autosetup": "Run auto-setup?"
+    "use_r": "Use R?"
```
I think we should still have a prompt here for clarity?

HarrisonWilde reviewed (commented):
Looks great overall, I have not had chance to properly test it yet. I am wondering if it's worth maintaining a sort of bare bones version that does what the cookie cutter used to do without the auto setup flag being raised? In most cases I don't see harm in making a venv + initialising git and pre-commit, did you consider some other early exit flag / keen to hear your thoughts on it and my other comments?

caldwellst committed:
Change approach to still allow no auto-setup

caldwellst committed:
Update arg name

caldwellst committed:
Address comment

caldwellst committed:
Remove duplicated strings in the postgen script

caldwellst committed:
Bail out if GH not setup yet

HarrisonWilde commented on docs/quickstart.md:
```diff
@@ -36,11 +36,14 @@ You will be prompted to enter the following information:
 -   `file_structure [standard]`: choose the complexity of your project. The options are: `simple` (basic and recommended for small projects: includes folders for analysis, getters and notebooks); `standard` (a good balance between simplicity and complexity: adds folders for utils, config and pipelines; moves notebooks into analysis); and `full` (adds folders separate from the module for documenta ...
 -   `use_r [no]`: whether you want to use R in your project. If you select `yes`, the cookiecutter will create an `renv` environment and set up some auxiliary things for the project, such as an `.Renviron` file for AWS credentials.
--   `autosetup [yes]`: this will automatically set up the project's virtual environment, pre-commit hooks and git repository for you. If you select `no`, you will have to do this manually later.'
+-   `auto_configure [yes]`: controls how much of the project configuration the cookiecutter performs automatically.
```
Perfect!

HarrisonWilde reviewed (approved):
LGTM, happy to merge and bump!

HarrisonWilde commented:
This will close #257 and #229 I will associate them with the PR
